pkg/generate/code/late_initialize.go

```diff
@@ -188,3 +182,146 @@ func LateInitializeFromReadOne(
+//			requeueDelay = int(math.Max(float64(requeueDelay), float64(delay)))
+//			incompleteInitialization = true
+//		}
+//	}
+//	if koWithDefaults.Spec.structA != nil {
+//		if koWithDefaults.Spec.structA.mapB != nil {
+//			if koWithDefaults.Spec.structA.mapB["structC"] != nil {
+//				if koWithDefaults.Spec.structA.mapB["structC"].valueD == nil {
+//					delay := (&acktypes.LateInitializationRetryConfig{MinBackoffSeconds:20, MaxBackoffSeconds: 0,}).GetExponentialBackoffSeconds(numInitAttempt)
+//					requeueDelay = int(math.Max(float64(requeueDelay), float64(delay)))
+//					incompleteInitialization = true
+//				}
+//			}
+//		}
+//	}
+func CalculateRequeueDelay(
+	cfg *ackgenconfig.Config,
+	r *model.CRD,
+	koVarName string,
+	initAttemptVarName string,
+	delayVarNameInt string,
+	incompleteInitializationVarNameBool string,
+	// Number of levels of indentation to use
+	indentLevel int,
+) string {
+	out := ""
+	sortedLateInitFieldNames, fieldNameToLateInitConfig := getSortedLateInitFieldsAndConfig(cfg, r)
+	for _, fName := range sortedLateInitFieldNames {
+		// split the field name by period
+		// each substring represents a field.
+		fNameParts := strings.Split(fName, ".")
+		// fNameIndentLevel tracks the indentation level for every new line added
+		// This variable is incremented when building nested if blocks and decremented when closing those if blocks.
+		fNameIndentLevel := indentLevel
+		// fParentPath keeps track of parent path for any fNamePart
+		fParentPath := ""
+		mapShapedParent := false
+		for i, fNamePart := range fNameParts {
+			if fNamePart == "" {
+				mapShapedParent = true
+				continue
+			}
+			indent := strings.Repeat("\t", fNameIndentLevel)
+			fNamePartAccesor := fmt.Sprintf("Spec%s.%s", fParentPath, fNamePart)
+			if mapShapedParent {
+				fNamePartAccesor = fmt.Sprintf("Spec%s[%q]", fParentPath, fNamePart)
+			}
+			// Handling for all parts except last one
+			if i != len(fNameParts)-1 {
+				out += fmt.Sprintf("%sif %s.%s != nil {\n", indent, koVarName, fNamePartAccesor)
+				// update fParentPath and fNameIndentLevel for next iteration
+				if mapShapedParent {
+					fParentPath = fmt.Sprintf("%s[%q]", fParentPath, fNamePart)
+					mapShapedParent = false
+				} else {
+					fParentPath = fmt.Sprintf("%s.%s", fParentPath, fNamePart)
+				}
+				fNameIndentLevel = fNameIndentLevel + 1
+			} else {
+				// handle last part here
+				// for last part, if the late initialized field is still nil, calculate the retry backoff using
+				// acktypes.LateInitializationRetryConfig abstraction and set the incompleteInitialization flag to true
+				out += fmt.Sprintf("%sif %s.%s == nil {\n", indent, koVarName, fNamePartAccesor)
+				fNameIndentLevel = fNameIndentLevel + 1
+				indent = strings.Repeat("\t", fNameIndentLevel)
+				minBackoffSeconds := fieldNameToLateInitConfig[fName].MinBackoffSeconds
+				maxBackoffSeconds := fieldNameToLateInitConfig[fName].MaxBackoffSeconds
```

Comment:
These are optional fields? Want to use a getter that provides a sane default?

Reply:
Yes, these are optional. I think 0 which is current default is the sane value already.
 if MinBackoffSeconds is 0, Exponential backoff will be 0, 1, 2, 4, 6, 8, 10 seconds..... which is fine.
And if MaxBackoffSeconds is 0, it is treated as user does not want a MaxBackoff limit. 

I can think of having a some sane upper bound for MaxBackoff but I do not think it is necessary. LMK your thoughts.

Reply:
Actually, I'm confused as to how we are supposed to support MaxBackoff? At the moment we are taking the max of each of the backoffs to determine the next delay, but that guarantees the backoff will be greater than or equal to every other field's backoff.
For instance, if we have a maxbackoff of 3, 4 and 5 - after some iterations we will reach the backoff of 5. This will exceed the maxbackoff of the 3 and 4. 

I would be fine with continuing the exponential backoff indefinitely, or hardcoding a limit (say 30s) that applies for every late initialized field.

Reply:
[Here](https://github.com/aws-controllers-k8s/runtime/pull/37/files/54699a88f8e14c2f7daace83b98490a9741904be#diff-f411262cb2fadbaec50818511b49921468a6fc23eb6a2bd9f3801294149b2102) is the implementation for calculating the backoff. 

Currently, I won't add a hardcoded maximum but if the need arises for it, We can introduce it. It is not a one way door.